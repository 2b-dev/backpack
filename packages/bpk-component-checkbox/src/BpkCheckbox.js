import React, { PropTypes } from 'react'

import './bpk-checkbox.scss'

const BpkCheckbox = ({ name, label, value, id, checked, onChange, disabled }) => {
  const classNames = [ 'bpk-checkbox' ]

  disabled ? classNames.push('bpk-checkbox--disabled') : null

  return (
    <label className={classNames.join(' ')}>
      <input
        type='checkbox'
        className='bpk-checkbox__input'
        name={name}
        value={value || name}
        id={id}
        checked={checked}
        onChange={onChange}
        disabled={disabled}
      />
      {label ? ` ${label}` : null}
    </label>
  )
}

BpkCheckbox.propTypes = {
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  id: PropTypes.string,
  checked: PropTypes.bool,
  value: PropTypes.string,
  label: PropTypes.string,
  disabled: PropTypes.bool
}

BpkCheckbox.defaultProps = {
  id: null,
  checked: false,
  value: null,
  label: null,
  disabled: false
}

export default BpkCheckbox
