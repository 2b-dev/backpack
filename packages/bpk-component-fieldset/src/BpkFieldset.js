import React, { cloneElement, PropTypes } from 'react';
import BpkLabel from 'bpk-component-label';
import BpkFormValidation from 'bpk-component-form-validation';

import './bpk-fieldset.scss';

const BpkFieldset = (props) => {
  const {
    children,
    label,
    className,
    validationMessage,
    valid,
    required,
    isCheckbox,
    validationProps,
    ...rest
  } = props;

  if (!children) {
    return null;
  }

  const classNames = ['bpk-fieldset'];
  const validationMessageId = `${children.props.id}_validation_message`;

  // Explicit check for false primitive value as undefined is
  // treated as neither valid nor invalid
  const isInvalid = isCheckbox
    ? valid === false
    : children.props.valid === false;

  const childrenProps = {
    'aria-required': required,
  };

  if (isCheckbox && required) {
    childrenProps.required = true;
  }

  if (validationMessage && isInvalid) {
    childrenProps['aria-describedby'] = validationMessageId;
  }

  const clonedChildren = cloneElement(children, childrenProps);

  if (className) { classNames.push(className); }

  return (
    <fieldset className={classNames.join(' ')} {...rest}>
      {!isCheckbox && (
        <BpkLabel htmlFor={children.props.id} required={required}>{label}</BpkLabel>
      )}
      {clonedChildren}
      {validationMessage && (
        <BpkFormValidation
          id={validationMessageId}
          expanded={isInvalid}
          isCheckbox={isCheckbox}
          {...validationProps}
        >
          {validationMessage}
        </BpkFormValidation>
      )}
    </fieldset>
  );
};

BpkFieldset.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.string,
  valid: PropTypes.bool,
  required: PropTypes.bool,
  className: PropTypes.string,
  validationMessage: PropTypes.string,
  isCheckbox: PropTypes.bool,
  validationProps: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

BpkFieldset.defaultProps = {
  label: null,
  valid: null,
  required: false,
  className: null,
  validationMessage: null,
  isCheckbox: false,
  validationProps: {},
};

export default BpkFieldset;

