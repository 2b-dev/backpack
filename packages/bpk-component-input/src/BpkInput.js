import React, { PropTypes } from 'react'

import './bpk-input.scss'

export const INPUT_TYPES = {
  TEXT: 'text',
  EMAIL: 'email',
  NUMBER: 'number',
  PASSWORD: 'password',
  TEL: 'tel'
}

const BpkInput = ({ name, value, onChange, placeholder, valid, type }) => {
  const classNames = [ 'bpk-input' ]

  valid
    ? classNames.push('bpk-input--valid')
    : valid === false ? classNames.push('bpk-input--invalid') : null

  return (
    <input
      className={classNames.join(' ')}
      name={name}
      value={value}
      onChange={onChange}
      placeholder={placeholder}
      type={type}
    />
  )
}

BpkInput.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  valid: PropTypes.bool,
  type: PropTypes.oneOf([
    INPUT_TYPES.TEXT,
    INPUT_TYPES.EMAIL,
    INPUT_TYPES.NUMBER,
    INPUT_TYPES.PASSWORD,
    INPUT_TYPES.TEL
  ])
}

BpkInput.defaultProps = {
  type: INPUT_TYPES.TEXT,
  placeholder: null,
  valid: null
}

export default BpkInput
