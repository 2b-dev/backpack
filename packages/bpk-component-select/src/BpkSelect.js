import React, { PropTypes } from 'react';

import './bpk-select.scss';

const BpkSelect = (props) => {
  const classNames = ['bpk-select'];
  const { valid, large, docked, dockedFirst, dockedMiddle, dockedLast, className, ...rest } = props;

  // Explicit check for false primitive value as undefined is
  // treated as neither valid nor invalid
  const isInvalid = valid === false;

  if (large) { classNames.push('bpk-select--large'); }
  if (docked) { classNames.push('bpk-select--docked'); }
  if (dockedFirst) { classNames.push('bpk-select--docked-first'); }
  if (dockedMiddle) { classNames.push('bpk-select--docked-middle'); }
  if (dockedLast) { classNames.push('bpk-select--docked-last'); }
  if (isInvalid) { classNames.push('bpk-select--invalid'); }
  if (className) { classNames.push(className); }

  return (
    <select className={classNames.join(' ')} aria-invalid={isInvalid} {...rest} />
  );
};

BpkSelect.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  className: PropTypes.string,
  valid: PropTypes.bool,
  large: PropTypes.bool,
  docked: PropTypes.bool,
  dockedFirst: PropTypes.bool,
  dockedMiddle: PropTypes.bool,
  dockedLast: PropTypes.bool,
};

BpkSelect.defaultProps = {
  className: null,
  valid: null,
  large: false,
  docked: false,
  dockedFirst: false,
  dockedMiddle: false,
  dockedLast: false,
};

export default BpkSelect;
