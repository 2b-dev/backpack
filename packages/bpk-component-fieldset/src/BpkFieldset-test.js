import React from 'react';
import renderer from 'react-test-renderer';
import BpkSelect from 'bpk-component-select';
import BpkCheckbox from 'bpk-component-checkbox';
import BpkInput, { INPUT_TYPES } from 'bpk-component-input';

import BpkFieldset from './BpkFieldset';

describe('BpkFieldset', () => {
  it('should render correctly with input component', () => {
    const tree = renderer.create(
      <BpkFieldset
        label="Name"
        validationMessage="Please enter a name"
      >
        <BpkInput
          id="name_input"
          name="name"
          type={INPUT_TYPES.TEXT}
          placeholder="e.g. Joe Bloggs"
          value=""
        />
      </BpkFieldset>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with input component and "valid" attribute', () => {
    const tree = renderer.create(
      <BpkFieldset
        label="Name"
        validationMessage="Please enter a name"
      >
        <BpkInput
          id="name_input"
          name="name"
          type={INPUT_TYPES.TEXT}
          placeholder="e.g. Joe Bloggs"
          value=""
          valid
        />
      </BpkFieldset>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with input component and "valid" attribute equal to false', () => {
    const tree = renderer.create(
      <BpkFieldset
        label="Name"
        validationMessage="Please enter a name"
      >
        <BpkInput
          id="name_input"
          name="name"
          type={INPUT_TYPES.TEXT}
          placeholder="e.g. Joe Bloggs"
          value=""
          valid={false}
        />
      </BpkFieldset>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with select component', () => {
    const tree = renderer.create(
      <BpkFieldset
        label="Fruits"
        validationMessage="Please select a fruit"
      >
        <BpkSelect
          id="fruits_select"
          name="fruits"
          value=""
        >
          <option value="">Please select...</option>
          <option value="apples">Apples</option>
          <option value="oranges">Oranges</option>
          <option value="pears">Pears</option>
          <option value="tomato" disabled>Tomato</option>
        </BpkSelect>
      </BpkFieldset>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with checkbox component', () => {
    const tree = renderer.create(
      <BpkFieldset
        isCheckbox
        validationMessage="You must accept the terms and conditions to continue"
      >
        <BpkCheckbox
          id="terms_and_conditions_checkbox"
          name="terms_and_conditions"
          label="I accept the terms and conditions"
        />
      </BpkFieldset>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
