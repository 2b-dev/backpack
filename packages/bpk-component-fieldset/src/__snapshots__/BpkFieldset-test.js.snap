exports[`BpkFieldset should render correctly with checkbox component 1`] = `
<fieldset
  className="bpk-fieldset">
  <label
    className="bpk-checkbox">
    <input
      className="bpk-checkbox__input"
      disabled={false}
      id="terms_and_conditions_checkbox"
      name="terms_and_conditions"
      type="checkbox" />
    I accept the terms and conditions
  </label>
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": 0,
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }>
    <div>
      <div
        className="bpk-form-validation__container">
        <div
          className="bpk-form-validation bpk-form-validation--is-checkbox"
          id="terms_and_conditions_checkbox_validation_message">
          You must accept the terms and conditions to continue
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`BpkFieldset should render correctly with input component 1`] = `
<fieldset
  className="bpk-fieldset">
  <label
    className="bpk-label"
    htmlFor="name_input">
    Name
  </label>
  <input
    aria-invalid={false}
    className="bpk-input"
    id="name_input"
    name="name"
    placeholder="e.g. Joe Bloggs"
    type="text"
    value="" />
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": 0,
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }>
    <div>
      <div
        className="bpk-form-validation__container">
        <div
          className="bpk-form-validation"
          id="name_input_validation_message">
          Please enter a name
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`BpkFieldset should render correctly with input component and "valid" attribute 1`] = `
<fieldset
  className="bpk-fieldset">
  <label
    className="bpk-label"
    htmlFor="name_input">
    Name
  </label>
  <input
    aria-invalid={false}
    className="bpk-input bpk-input--valid"
    id="name_input"
    name="name"
    placeholder="e.g. Joe Bloggs"
    type="text"
    value="" />
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": 0,
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }>
    <div>
      <div
        className="bpk-form-validation__container">
        <div
          className="bpk-form-validation"
          id="name_input_validation_message">
          Please enter a name
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`BpkFieldset should render correctly with input component and "valid" attribute equal to false 1`] = `
<fieldset
  className="bpk-fieldset">
  <label
    className="bpk-label"
    htmlFor="name_input">
    Name
  </label>
  <input
    aria-describedby="name_input_validation_message"
    aria-invalid={true}
    className="bpk-input bpk-input--invalid"
    id="name_input"
    name="name"
    placeholder="e.g. Joe Bloggs"
    type="text"
    value="" />
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": "auto",
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }>
    <div>
      <div
        className="bpk-form-validation__container">
        <div
          className="bpk-form-validation bpk-form-validation--appear"
          id="name_input_validation_message">
          Please enter a name
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`BpkFieldset should render correctly with select component 1`] = `
<fieldset
  className="bpk-fieldset">
  <label
    className="bpk-label"
    htmlFor="fruits_select">
    Fruits
  </label>
  <select
    aria-invalid={false}
    className="bpk-select"
    id="fruits_select"
    name="fruits"
    value="">
    <option
      value="">
      Please select...
    </option>
    <option
      value="apples">
      Apples
    </option>
    <option
      value="oranges">
      Oranges
    </option>
    <option
      value="pears">
      Pears
    </option>
    <option
      disabled={true}
      value="tomato">
      Tomato
    </option>
  </select>
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": 0,
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }>
    <div>
      <div
        className="bpk-form-validation__container">
        <div
          className="bpk-form-validation"
          id="fruits_select_validation_message">
          Please select a fruit
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
